mlflow/utils/environment.py

```diff
@@ -18,6 +25,91 @@
 _CONDA_ENV_FILE_NAME = "conda.yaml"
 _REQUIREMENTS_FILE_NAME = "requirements.txt"
 _CONSTRAINTS_FILE_NAME = "constraints.txt"
+_PYTHON_ENV_FILE_NAME = "python-env.yaml"
+
+
+class EnvManager(Enum):
+    LOCAL = "local"
+    CONDA = "conda"
+    VIRTUALENV = "virtualenv"
+
+
+_PYTHON_REQUIREMENT_REGEX = re.compile(r"^python=([\d.]+)$")
+
+
+class PythonEnv:
+    def __init__(self, python=None, build_dependencies=None, dependencies=None):
+        self.python = python
+        self.build_dependencies = build_dependencies
+        self.dependencies = dependencies
+
+    def __str__(self):
+        return str(self.to_dict())
+
+    @staticmethod
+    def get_current_python():
+        return ".".join(map(str, sys.version_info[:3]))
+
+    @staticmethod
+    def _get_package_version(package_name):
+        try:
+            return __import__(package_name).__version__
+        except (ImportError, AttributeError):
+            return None
+
+    @staticmethod
+    def get_default_build_dependencies():
+        build_dependencies = []
+        for package in ["pip", "setuptools", "wheel"]:
+            version = PythonEnv._get_package_version(package)
+            dep = (package + "==" + version) if version else package
+            build_dependencies.append(dep)
+        return build_dependencies
+
+    def to_dict(self):
+        return self.__dict__.copy()
+
+    @classmethod
+    def from_dict(cls, dct):
+        return cls(**dct)
+
+    def to_yaml(self, path):
```

Comment:
Nit: maybe the function name "save_as_yaml" is more clear. "to_yaml" looks like a function returning a yaml file content.

Reply:
https://github.com/mlflow/mlflow/blob/dab955c02f762852a0e4ec8cf01c7baaffdf598d/mlflow/models/model.py#L192-L195

I chose `to_yaml` because `mlflow.models.Model` uses the same name.

Reply:
pandas also uses this pattern.